### ISSUE TITLE
Conversion funnel analytics (contact, careers, newsletter, solution pages)

### ISSUE THREAD

**Issue by lizhsr** [SimisRnD/simis-cms#565]
Split out of #479 (item 6) during scoping -- see the [full investigation](https://github.com/SimisRnD/simis-cms/issues/479) for how all 11 items were verified against actual code before splitting.

Track lead/user progression through key workflows (contact form, career application, newsletter signup, solution-page engagement), with conversion rate and drop-off at each stage.

## Investigation notes (2026-07-28)

Confirmed with no ambiguity: **zero funnel/attribution infrastructure exists in this codebase.** No hits for "funnel" as a concept anywhere (only an unrelated code comment about encryption). No hits for "attribution". No stage/step/event-sequence table in any install or upgrade migration. Every existing "conversion" reference in the codebase is about file-format conversion (CSV, markdown), not marketing funnels.

This is a genuinely large, multi-phase feature -- it needs a real event-sequence/attribution data model built from scratch (what counts as a "stage", how a visitor is tracked across stages, how drop-off is computed), not an extension of an existing table. Recommend treating this as its own scoped design effort rather than a single PR.

## Required
1. Contact form funnel: page view → submission → follow-up sent → response
2. Career application: landing page → application → submission → HR review
3. Newsletter signup: page view → form → email confirmation → active subscriber
4. Solution page engagement: product page → contact form → sales follow-up

For each: conversion rate per stage, drop-off %, avg time between stages, best-converting traffic sources, geographic/device breakdown, identify leaky stages.

## Acceptance Criteria
- Can see where leads get lost and prioritize fixes

---

## Scoping (updated 2026-07-29)

**Correction to the note above:** "every existing 'conversion' reference is about file-format conversion" is no longer accurate — issue #563 landed the same day this was written and added a real (if narrow) page-view-to-form-submission conversion-rate tile. Doesn't change the "no funnel infrastructure" conclusion, but it's real prior art worth building on rather than a false negative to leave uncorrected.

### What already exists to build on

- **A working cross-table join key**: `sessions.session_id` = `web_page_hits.session_id` = `form_data.session_id` = `emails.session_id` (all `VARCHAR(255)`, none FK'd, but consistently populated from the same `UserSession.getSessionId()` / servlet `HttpSession` ID). Reliable for same-session sequencing (60-minute servlet session timeout) — a page view now and a form submission or newsletter signup minutes later already correlate with zero schema changes.
- **`web_page_hits.web_page_id`** is populated today (`SaveWebPageHitCommand`, gets it from `PageServlet`'s already-loaded `WebPage`). The page-view side of every funnel is solid.
- **Geo columns** on `sessions` (continent/country/city/state/lat/long/timezone), joinable via `session_id` — with the existing caveat that only non-anonymous (logged-in) sessions get city/postal/lat-long; anonymous visitors get country/state/continent/timezone only.
- **The `generate_series` + `LEFT JOIN` zero-filled time-bucketing pattern**, used identically in four places already (`UserRepository.findMonthlyUserRegistrations`, `MailingListMemberRepository.findMonthlySubscriptions`, `FormDataRepository.findDailySubmissions`, `SessionRepository.findDailySessionsByBotStatus`) — the exact shape a funnel trend chart should mirror, down to returning `StatisticsData{label, value}`.
- **#563's `WebPageHitRepository.countPageViews(pagePath, start, end)`** and **`FormDataRepository.countSubmissions(formUniqueId, start, end)`** — real, tested, bot-excluded stage-count primitives already wired into `SiteStatsWidget`'s `conversion-rate` report. That report's own code comment states the exact gap this issue needs to close: *"there's no reliable automatic link between a page and the form(s) embedded on it."*

### Real gaps (not just "no funnel table" — specific, load-bearing ones)

- **No `web_page_id` on `form_data` or `emails`.** The page-view side captures it; the submission/signup side doesn't. This is the single most important gap to close — without it, "which page did this conversion come from" stays a hand-configured pairing forever, exactly as #563's tile already is.
- **No referrer on `form_data` at all**, and `sessions.referer` is captured **once, at session start** — it's the visit's entry referrer, not "the page the visitor was on when they converted." Good enough for "which traffic source drove this visit," not for per-page-view referrer chains.
- **Cross-day correlation is structurally limited, not just unbuilt.** The only durable cross-session identifier is the `visitors.token` cookie (1 year), but (a) `web_page_hits` has no `visitor_id` column at all, and (b) when `analytics.cookieless` is on, the visitor identifier is a hash salted per-calendar-day and never persisted — by design, unrecoverable across days. A funnel stage that completes more than a session (60 min) after the previous one is fundamentally not always traceable to the same visitor under this site's own privacy settings, not a gap this issue can code around.
- **No UTM capture, no device/browser parsing.** Only a raw `Referer` header and a raw `user_agent` string exist anywhere. Per your call: ship v1 without these (referrer *host* as a coarse traffic source, no device breakdown) rather than adding a new UA-parsing dependency now.
- **Retention is asymmetric today**: `web_page_hits` rows are hard-deleted past `analytics.retentionDays`; `sessions` rows are PII-scrubbed but kept. A new funnel-events table needs its own explicit retention decision, not an assumption that the source tables will still be there.

### Scope decisions (confirmed 2026-07-29)

- **Career-application funnel: out of scope.** Applications go through HR Jazz, an external system this codebase doesn't touch — there's nothing here to instrument beyond, optionally, "careers page viewed → clicked through to HR Jazz" as a cheap 2-stage bonus using the same page-view tracking. Not required; add later if wanted.
- **Solution-page engagement: not a fourth funnel.** It's the *same* page → contact-form → processed mechanism as the contact-form funnel below, just scoped to a different set of pages (and optionally a different form). No CRM/Lead entity — "sales follow-up" is the same `processed` proxy as the contact funnel's stage 3, just filtered to solution-page submissions. Comes for free as a second named funnel definition once the contact-form funnel exists, not as separate infrastructure.
- **Newsletter signup funnel: 3 real stages, not 4.** `page view → form submitted → active subscriber`. "Email confirmation" doesn't exist anywhere in this codebase — there's no double opt-in; a signup is `is_valid = true` in the same instant it's inserted. Dropped from the funnel rather than built as prerequisite work.
- **Traffic-source / geo / device**: ship with what already exists (referrer host, session geo columns with the anonymous-visitor caveat above). No UTM capture, no device-type breakdown, in this pass.
- **Contact-form "follow-up sent → response"**: only stages 1–2 (view → submission) get real, reliable tracking. `form_data.processed` (an admin's single-click triage flag, not proof a reply was sent) is used as an approximate, explicitly-caveated stage 3. Stage 4 (customer response) is dropped — nothing in this codebase can observe an outbound reply today, and building that is a distinct, separate feature (real reply-tracking/logging), not part of this issue.

### Proposed data model

One new table, deliberately denormalized (matching how every other reporting concept in this codebase — job names, report keys — is a set of string constants, not a driven-by-schema registry):

```sql
CREATE TABLE funnel_events (
  funnel_event_id BIGSERIAL PRIMARY KEY,
  funnel_name VARCHAR(50) NOT NULL,      -- 'contact_form', 'newsletter_signup', 'solution_page_<name>', ...
  stage_name VARCHAR(50) NOT NULL,       -- 'page_view', 'form_submitted', 'processed', 'active_subscriber'
  stage_order SMALLINT NOT NULL,         -- 1, 2, 3... within this funnel; lets drop-off math avoid a lookup join
  session_id VARCHAR(255),
  web_page_id BIGINT REFERENCES web_pages(web_page_id),
  referer VARCHAR(255),
  occurred TIMESTAMP(3) DEFAULT CURRENT_TIMESTAMP NOT NULL
);
CREATE INDEX funnel_events_funnel_stage_idx ON funnel_events(funnel_name, stage_name);
CREATE INDEX funnel_events_session_idx ON funnel_events(session_id);
CREATE INDEX funnel_events_occurred_idx ON funnel_events(occurred);
```

Which page paths / form-unique-IDs count as which funnel's stages is configuration (mirroring the existing `pagePath`/`formUniqueId` widget-preference idiom #563 already established), not new schema — a "solution page" funnel is just a different config pointed at the same event table and the same instrumentation.

### Instrumentation points

- **`PageServlet`**: already has the loaded `WebPage` in scope for every request (used for draft/publish/redirect checks) — record a `page_view` event when the current path matches a configured funnel, right alongside the existing `SaveWebPageHitCommand.saveHit(...)` call.
- **`FormWidget.post()`**: record `form_submitted` when `formUniqueId` matches a configured funnel. Requires first closing the `web_page_id`-on-`form_data` gap above (thread the page's id through `WidgetContext`, the same object `PageServlet` already has it on — currently not passed down at all).
- **`FormDataListWidget`'s "Mark as Processed" handler**: record `processed`, reusing the *original* submission's stored `session_id`/`web_page_id` — this event fires days later from an admin's own session, so it must look those up from the `form_data` row, not from the admin's current request.
- **`EmailSubscribeWidget.post()` / `EmailSubscribeAjax.execute()`** (both funnel into `SaveEmailCommand.saveEmail()`): record `form_submitted` and `active_subscriber` together, since both happen in the same synchronous call today. Open detail to settle during implementation: the signup widget is commonly embedded site-wide (e.g. the footer), so "stage 1 = page view" may need to mean "any page view earlier in the visit," not one dedicated landing page the way the contact-form funnel's stage 1 does.

### Reporting

Per-funnel stage counts (via `countPageViews`/`countSubmissions`-style primitives against `funnel_events` grouped by `stage_name`) give conversion rate and drop-off directly; `occurred` timestamps diffed per `session_id` across consecutive stages give avg time-between-stages; the `generate_series` pattern gives trend-over-time charts; referrer host and the `sessions` geo join give the two breakdowns being shipped in v1. All feed the existing `StatisticsData`/`SiteStatsWidget` reporting path new funnel-shaped report types, the same way every other dashboard metric in this codebase is added.

### Proposed phasing (separate PRs, not one)

1. **`funnel_events` table + write path + contact-form funnel** (view → submitted → processed) — the real infrastructure investment; everything after this is mostly configuration.
2. **Newsletter signup funnel** (view → submitted → active subscriber), reusing phase 1's table and write path.
3. **Solution-page engagement** as a second contact-form-shaped funnel definition — should be small, since it's phase 1's mechanism with different config.
4. *(Optional, not required)* Careers page-view → HR Jazz click-through, if wanted later.

## Required (revised scope)
1. `funnel_events` table (install + upgrade migration) and a thin write path
2. Contact-form funnel: page view → form submission → processed (caveated), with per-stage conversion rate, drop-off %, avg time-between-stages, referrer-host and geo breakdown
3. Newsletter signup funnel: page view → form submission → active subscriber, same reporting shape
4. Solution-page engagement reported as its own funnel view, reusing #2's mechanism against a different page/form scope
5. `web_page_id` capture added to `form_data` (and threaded through `WidgetContext`), closing the gap blocking #563's tile from ever being fully automatic

**Explicitly out of scope for this issue:** career-application funnel (external system — HR Jazz); UTM capture; device/browser-type breakdown; email double opt-in / confirmation; real reply/response tracking for contact-form follow-up; any CRM/Lead entity.

## Acceptance Criteria (revised)
- Admins can see conversion rate, drop-off %, and avg time-between-stages for the contact-form and newsletter-signup funnels, with referrer-host and geographic breakdown per stage
- Solution-page engagement is visible as its own funnel view built on the same mechanism
- Every stage definition and every explicitly-dropped/caveated stage (see Scope decisions above) is documented in-product, not just here, so an admin reading the dashboard understands what "processed" does and doesn't guarantee

**Comment by lizhsr** [comment]
Slice 1 shipped in #892 — view/submitted/processed funnel tracking for the contact form, gated behind two site properties (blank by default, so it's inert until configured), with a per-stage/drop-off report on /admin/community/analytics.

A pre-ship adversarial review caught and fixed 5 issues before opening the PR, most notably a fatal XML syntax error that would have broken every admin page's layout load — see the PR description for the full list.

Leaving this open for any follow-up funnels (newsletter signup, solution-page engagement) the table was designed to support without a schema change.
